qa/rpc-tests/p2p-segwit.py

```diff
@@ -0,0 +1,1575 @@
+        self.test_node.test_witness_block(block, accepted=True)
+
+        self.utxo.pop()
+        self.utxo.append(UTXO(tx2.sha256, 0, tx2.vout[0].nValue))
+
+
+    def test_witness_tx_relay_before_segwit_activation(self):
+        print("\tTesting relay of witness transactions")
+        # Generate a transaction that doesn't require a witness, but send it
+        # with a witness.  Should be rejected for premature-witness, but should
+        # not be added to recently rejected list.
+        assert(len(self.utxo))
+        tx = CTransaction()
+        tx.vin.append(CTxIn(COutPoint(self.utxo[0].sha256, self.utxo[0].n), b""))
+        tx.vout.append(CTxOut(self.utxo[0].nValue-1000, CScript([OP_TRUE])))
+        tx.wit.vtxinwit.append(CTxinWitness())
+        tx.wit.vtxinwit[0].scriptWitness.stack = [ b'a' ]
+        tx.rehash()
+
+        tx_hash = tx.sha256
+        tx_value = tx.vout[0].nValue
+
+        # Verify that if a peer doesn't set nServices to include NODE_WITNESS,
+        # the getdata is just for the non-witness portion.
+        self.old_node.announce_tx_and_wait_for_getdata(tx)
+        assert(self.old_node.last_getdata.inv[0].type == 1)
+
+        # Since we haven't delivered the tx yet, inv'ing the same tx from
+        # a witness transaction ought not result in a getdata.
+        try:
+            self.test_node.announce_tx_and_wait_for_getdata(tx, timeout=2)
+            print("Error: duplicate tx getdata!")
+            assert(False)
+        except AssertionError as e:
+            pass
+
+        # Delivering this transaction with witness should fail (no matter who
+        # its from)
+        assert_equal(len(self.nodes[0].getrawmempool()), 0)
+        assert_equal(len(self.nodes[1].getrawmempool()), 0)
+        self.old_node.test_transaction_acceptance(tx, with_witness=True, accepted=False)
+        self.test_node.test_transaction_acceptance(tx, with_witness=True, accepted=False)
+
+        # But eliminating the witness should fix it
+        self.test_node.test_transaction_acceptance(tx, with_witness=False, accepted=True)
+
+        # Verify that inv's to test_node come with getdata's for witness tx's
+        # Just tweak the transaction, announce it, and verify we get a getdata
+        # for a witness_tx
+        tx.vout[0].scriptPubKey = CScript([OP_TRUE, OP_TRUE])
+        tx.rehash()
+        self.test_node.announce_tx_and_wait_for_getdata(tx)
+        assert(self.test_node.last_getdata.inv[0].type == 1|MSG_WITNESS_FLAG)
+
+        # Cleanup: mine the first transaction and update utxo
+        self.nodes[0].generate(1)
+        assert_equal(len(self.nodes[0].getrawmempool()),  0)
+
+        self.utxo.pop(0)
+        self.utxo.append(UTXO(tx_hash, 0, tx_value))
+
+
+    # After segwit activates, verify that mempool:
+    # - rejects transactions with unnecessary/extra witnesses
+    # - accepts transactions with valid witnesses
+    # and that witness transactions are relayed to non-upgraded peers.
```

Comment:
Good observation; we should verify that sending the two different kinds of `getdata` give the expected results.